### ISSUE TITLE
PyInstaller-made app executable size balloons 10x when going from poetry 1.1 to poetry 1.3 or poetry 1.5 

### ISSUE THREAD

**Issue by hintron** [pyinstaller/pyinstaller#7856]
### Context information

* Output of `pyinstaller --version`: ```5.13.0```
* Version of Python: 3.10.12
* Platform: Linux - SUSE 12.5
* How you installed Python: Conda w/ conda forge
* Did you also try this on another platform? Does it work there? --> IIRC I replicated this issue on my local Windows machine back in January, so it seems OS-independent.
* Version of poetry: 1.3 and/or 1.5

### Make sure [everything is packaged correctly](https://github.com/pyinstaller/pyinstaller/wiki/How-to-Report-Bugs#make-sure-everything-is-packaged-correctly)

  * [x] start with clean installation
  * [ ] use the latest development version
  * [x] Run your frozen program **from a command window (shell)** — instead of double-clicking on it
  * [ ] Package your program in **--onedir mode**
  * [ ] Package **without UPX**, say: use the option `--noupx` or set `upx=False` in your .spec-file
  * [ ] Repackage you application in **verbose/debug mode**. For this, pass the option `--debug` to `pyi-makespec` or `pyinstaller` or use `EXE(..., debug=1, ...)` in your .spec file.

### A minimal example program which shows the error

Basically, with poetry 1.1.5, I am able to make a 50 MB executable using the latest released PyInstaller and this [`pyproject.toml`][1]. This works as expected. However, when I upgrade poetry to 1.3 or 1.5, PyInstaller picks up _way_ more stuff, to the point that the executable balloons up to 458 MB.

(Here is the issue I opened with Poetry, in which they basically say "not our problem": https://github.com/python-poetry/poetry/issues/8323)

I'll provide more in-depth debug info as needed, assuming someone wants to look into this issue.

[1]: https://gist.github.com/hintron/f9b5f04a0052602caa9bc8db052d6500

**Comment by rokm** [comment]
PyInstaller does not properly support symbolic links, so building with Anaconda-installed `numpy` is bound to balloon your application size due to MKL shared libs being duplicated several times. It's a known issue that will be solved in v6.x with added symbolic link preservation support. So for now, avoid pyinstaller with Anaconda on Linux and macOS.

Regarding poetry 1.1 vs 1.3/1.5, my guess would be that they give you different build environments, i.e., install packages from different sources. Can you verify that? Are the versions of installed packages (`numpy`, `pandas`, etc.) the same in both build environments? Are in both cases packages installed from Anaconda, or did poetry 1.1 perhaps install `numpy` from PyPI via `pip`?

But you should also take a look at what was actually collected into the executable. Use `pyi-archive-viewer` utility if this is `onefile` build; although it might be easier to temporarily switch to `onedir` build, so you can directly inspect the contents.

**Comment by hintron** [comment]
> PyInstaller does not properly support symbolic links, so building with Anaconda-installed `numpy` is bound to balloon your application size due to MKL shared libs being duplicated several times. It's a known issue that will be solved in v6.x with added symbolic link preservation support. So for now, avoid pyinstaller with Anaconda on Linux and macOS.

Ok. I had numpy before and after changing poetry, but I haven't verified if it's coming from the same place. It seems that the newer versions of poetry download more Python wheels - perhaps that is playing a part.

> Regarding poetry 1.1 vs 1.3/1.5, my guess would be that they give you different build environments, i.e., install packages from different sources. Can you verify that? Are the versions of installed packages (`numpy`, `pandas`, etc.) the same in both build environments? Are in both cases packages installed from Anaconda, or did poetry 1.1 perhaps install `numpy` from PyPI via `pip`?

What's the best way to do that? Compare `poetry.lock` files? Or compare the package directories themselves?

> But you should also take a look at what was actually collected into the executable. Use `pyi-archive-viewer` utility if this is `onefile` build; although it might be easier to temporarily switch to `onedir` build, so you can directly inspect the contents.

So my PyInstaller command looks like this:
```
pyinstaller 
--noconfirm 
--add-data ../tlt-misc:tlt-misc 
--paths ../tlt-misc/scripts 
--hidden-import h5py 
--hidden-import json 
--hidden-import pandas 
--hidden-import python_jsonschema_objects 
--hidden-import tlt_scriptlib 
--hidden-import yaml 
-D ../.env/envs/my-app/bin/tlt
```

With poetry 1.1, this is the result:

```shell
$ du -sh dist/tlt-src/
148M    dist/tlt-src/
$ ls dist/tlt-src/
Markdown-3.4.4.dist-info                       jsonschema                     libhdf5_hl-408a1995.so.200.1.0       libuuid.so.1
_cffi_backend.cpython-310-x86_64-linux-gnu.so  jsonschema-4.18.4.dist-info    liblzma.so.5                         libz.so.1
attrs-23.1.0.dist-info                         jsonschema_specifications      libopenblas64_p-r0-7a851222.3.23.so  markupsafe
base_library.zip                               lib-dynload                    libpython3.10.so.1.0                 numpy
brotli                                         libaec-8db54129.so.0.0.10      libquadmath-96973f99.so.0.0.0        pandas
certifi                                        libbz2.so.1.0                  libreadline.so.8                     pytz
charset_normalizer                             libcrypto.so.3                 libsqlite3.so.0                      rpds
cryptography                                   libffi.so.8                    libssl.so.3                          tlt
cryptography-41.0.2.dist-info                  libgcc_s.so.1                  libstdc++.so.6                       tlt-misc
h5py                                           libgfortran-040039e1.so.5.0.0  libsz-60c3b967.so.2.0.1              yaml
importlib_metadata-1.7.0.dist-info             libhdf5-45610c27.so.200.2.0    libtinfow.so.6
```

With poetry 1.5, things ballooned:

```
$ du -sh dist-poetry-1.5/tlt-src/
1.5G    dist-poetry-1.5/tlt-src/
$ ls dist-poetry-1.5/tlt-src/
Markdown-3.4.4.dist-info                       libmenu.so.6                      libomptarget.rtl.x86_64.so
_cffi_backend.cpython-310-x86_64-linux-gnu.so  libmenu.so.6.4                    libomptarget.so
attrs-23.1.0.dist-info                         libmenuw.so                       libomptarget.sycl.wrap.so
base_library.zip                               libmenuw.so.6                     libopenblas64_p-r0-5007b62f.3.23.dev.so
brotli                                         libmenuw.so.6.4                   libpanel.so
certifi                                        libmkl_avx.so.2                   libpanel.so.6
charset_normalizer                             libmkl_avx2.so.2                  libpanel.so.6.4
cryptography                                   libmkl_avx512.so.2                libpanelw.so
cryptography-41.0.2.dist-info                  libmkl_blacs_intelmpi_ilp64.so    libpanelw.so.6
h5py                                           libmkl_blacs_intelmpi_ilp64.so.2  libpanelw.so.6.4
importlib_metadata-6.0.0.dist-info             libmkl_blacs_intelmpi_lp64.so     libpython3.10.so
jsonschema                                     libmkl_blacs_intelmpi_lp64.so.2   libpython3.10.so.1.0
jsonschema-4.17.3.dist-info                    libmkl_blacs_openmpi_ilp64.so     libpython3.so
lib-dynload                                    libmkl_blacs_openmpi_ilp64.so.2   libquadmath-96973f99.so.0.0.0
libaec-8db54129.so.0.0.10                      libmkl_blacs_openmpi_lp64.so      libquadmath.so
libasan.so                                     libmkl_blacs_openmpi_lp64.so.2    libquadmath.so.0
libasan.so.6                                   libmkl_cdft_core.so               libquadmath.so.0.0.0
libasan.so.6.0.0                               libmkl_cdft_core.so.2             libreadline.so
libatomic.so                                   libmkl_core.so                    libreadline.so.8
libatomic.so.1                                 libmkl_core.so.2                  libreadline.so.8.2
libatomic.so.1.2.0                             libmkl_def.so.2                   libsqlite3.so
libbz2.so                                      libmkl_gf_ilp64.so                libsqlite3.so.0
libbz2.so.1.0                                  libmkl_gf_ilp64.so.2              libsqlite3.so.0.8.6
libbz2.so.1.0.8                                libmkl_gf_lp64.so                 libssl.so
libcrypto.so                                   libmkl_gf_lp64.so.2               libssl.so.3
libcrypto.so.3                                 libmkl_gnu_thread.so              libstdc++.so
libffi.7.so                                    libmkl_gnu_thread.so.2            libstdc++.so.6
libffi.8.so                                    libmkl_intel_ilp64.so             libstdc++.so.6.0.29
libffi.so                                      libmkl_intel_ilp64.so.2           libsz-60c3b967.so.2.0.1
libffi.so.7                                    libmkl_intel_lp64.so              libtbb.so.12
libffi.so.8                                    libmkl_intel_lp64.so.2            libtbb.so.12.8
libffi.so.8.1.2                                libmkl_intel_thread.so            libtbbmalloc.so.2
libform.so                                     libmkl_intel_thread.so.2          libtbbmalloc.so.2.8
libform.so.6                                   libmkl_mc.so.2                    libtbbmalloc_proxy.so.2
libform.so.6.4                                 libmkl_mc3.so.2                   libtbbmalloc_proxy.so.2.8
libformw.so                                    libmkl_pgi_thread.so              libtcl8.6.so
libformw.so.6                                  libmkl_pgi_thread.so.2            libtinfo.so
libformw.so.6.4                                libmkl_rt.so                      libtinfo.so.6
libgcc_s.so                                    libmkl_rt.so.2                    libtinfo.so.6.4
libgcc_s.so.1                                  libmkl_scalapack_ilp64.so         libtinfow.so
libgfortran-040039e1.so.5.0.0                  libmkl_scalapack_ilp64.so.2       libtinfow.so.6
libgomp.so                                     libmkl_scalapack_lp64.so          libtinfow.so.6.4
libgomp.so.1                                   libmkl_scalapack_lp64.so.2        libtk8.6.so
libgomp.so.1.0.0                               libmkl_sequential.so              libtsan.so
libhdf5-45610c27.so.200.2.0                    libmkl_sequential.so.2            libtsan.so.0
libhdf5_hl-408a1995.so.200.1.0                 libmkl_tbb_thread.so              libtsan.so.0.0.0
libhistory.so                                  libmkl_tbb_thread.so.2            libubsan.so
libhistory.so.8                                libmkl_vml_avx.so.2               libubsan.so.1
libhistory.so.8.2                              libmkl_vml_avx2.so.2              libubsan.so.1.0.0
libiomp5.so                                    libmkl_vml_avx512.so.2            libuuid.so
libiomp5_db.so                                 libmkl_vml_cmpt.so.2              libuuid.so.1
libiompstubs5.so                               libmkl_vml_def.so.2               libuuid.so.1.3.0
libitm.so                                      libmkl_vml_mc.so.2                libz.so
libitm.so.1                                    libmkl_vml_mc3.so.2               libz.so.1
libitm.so.1.0.0                                libncurses.so                     libz.so.1.2.13
liblsan.so                                     libncurses.so.6                   markupsafe
liblsan.so.0                                   libncurses.so.6.4                 numpy
liblsan.so.0.0.0                               libncursesw.so                    pandas
liblzma.so                                     libncursesw.so.6                  pytz
liblzma.so.5                                   libncursesw.so.6.4                tlt
liblzma.so.5.4.2                               libomptarget.rtl.level0.so        tlt-misc
libmenu.so                                     libomptarget.rtl.opencl.so        yaml
```

Using `ncdu`, I can see what is taking up all the space:

1.1:
```
   31.7 MiB [##########]  libopenblas64_p-r0-7a851222.3.23.so
   20.1 MiB [######    ] /pandas
   16.7 MiB [#####     ]  libpython3.10.so.1.0
   12.2 MiB [###       ] /lib-dynload
   10.3 MiB [###       ] /numpy
    8.3 MiB [##        ]  tlt
    7.8 MiB [##        ] /h5py
    6.6 MiB [##        ] /cryptography
    4.9 MiB [#         ]  libcrypto.so.3
    4.7 MiB [#         ]  libhdf5-45610c27.so.200.2.0
    4.6 MiB [#         ] /rpds
    2.6 MiB [          ] /pytz
    2.6 MiB [          ]  libgfortran-040039e1.so.5.0.0
    2.5 MiB [          ] /brotli
    2.1 MiB [          ] /yaml
    2.1 MiB [          ]  libstdc++.so.6
    1.4 MiB [          ]  libsqlite3.so.0
    1.0 MiB [          ]  base_library.zip
  760.0 KiB [          ]  libssl.so.3
  564.0 KiB [          ] /tlt-misc
  472.0 KiB [          ]  libgcc_s.so.1
  444.0 KiB [          ] /charset_normalizer
  408.0 KiB [          ]  libreadline.so.8
  288.0 KiB [          ]  libtinfow.so.6
  284.0 KiB [          ] /certifi
  248.0 KiB [          ]  libquadmath-96973f99.so.0.0.0
  216.0 KiB [          ]  _cffi_backend.cpython-310-x86_64-linux-gnu.so
  216.0 KiB [          ]  liblzma.so.5
  180.0 KiB [          ]  libhdf5_hl-408a1995.so.200.1.0
  132.0 KiB [          ] /jsonschema
  132.0 KiB [          ] /jsonschema_specifications
  128.0 KiB [          ]  libz.so.1
  124.0 KiB [          ]  libaec-8db54129.so.0.0.10
   88.0 KiB [          ]  libbz2.so.1.0
   76.0 KiB [          ]  libffi.so.8
   56.0 KiB [          ] /cryptography-41.0.2.dist-info
   52.0 KiB [          ] /markupsafe
   44.0 KiB [          ]  libsz-60c3b967.so.2.0.1
   40.0 KiB [          ]  libuuid.so.1
   36.0 KiB [          ] /attrs-23.1.0.dist-info
   36.0 KiB [          ] /jsonschema-4.18.4.dist-info
   36.0 KiB [          ] /Markdown-3.4.4.dist-info
   24.0 KiB [          ] /importlib_metadata-1.7.0.dist-info 
```

1.5:
```
   72.0 MiB [##########]  libmkl_core.so.2                                                                                                
   72.0 MiB [##########]  libmkl_core.so
   64.5 MiB [########  ]  libmkl_avx512.so.2
   63.2 MiB [########  ]  libmkl_intel_thread.so.2
   63.2 MiB [########  ]  libmkl_intel_thread.so
   51.3 MiB [#######   ]  libmkl_avx.so.2
   48.7 MiB [######    ]  libmkl_mc3.so.2
   48.6 MiB [######    ]  libmkl_avx2.so.2
   47.2 MiB [######    ]  libmkl_mc.so.2
   41.6 MiB [#####     ]  libmkl_tbb_thread.so.2
   41.6 MiB [#####     ]  libmkl_tbb_thread.so
   41.2 MiB [#####     ]  libmkl_def.so.2
   37.3 MiB [#####     ]  libmkl_pgi_thread.so.2
   37.3 MiB [#####     ]  libmkl_pgi_thread.so
   33.6 MiB [####      ]  libopenblas64_p-r0-5007b62f.3.23.dev.so
   30.6 MiB [####      ]  libmkl_gnu_thread.so.2
   30.6 MiB [####      ]  libmkl_gnu_thread.so
   28.6 MiB [###       ]  libmkl_sequential.so.2
   28.6 MiB [###       ]  libmkl_sequential.so
   21.1 MiB [##        ]  libmkl_intel_lp64.so.2
   21.1 MiB [##        ]  libmkl_intel_lp64.so
   20.1 MiB [##        ] /pandas
   17.7 MiB [##        ]  libmkl_gf_lp64.so.2
   17.7 MiB [##        ]  libmkl_gf_lp64.so
   17.5 MiB [##        ]  libmkl_intel_ilp64.so.2
   17.5 MiB [##        ]  libmkl_intel_ilp64.so
   17.2 MiB [##        ]  libstdc++.so.6.0.29
   17.2 MiB [##        ]  libstdc++.so
   16.7 MiB [##        ]  libpython3.10.so.1.0
   16.7 MiB [##        ]  libpython3.10.so
   16.2 MiB [##        ]  libomptarget.rtl.level0.so
   15.4 MiB [##        ]  libmkl_vml_avx.so.2
   15.2 MiB [##        ]  libmkl_rt.so.2
   15.2 MiB [##        ]  libmkl_rt.so
   15.0 MiB [##        ]  libomptarget.rtl.opencl.so
   14.6 MiB [##        ]  libmkl_vml_avx2.so.2
   14.3 MiB [#         ]  libmkl_vml_mc.so.2
   14.2 MiB [#         ]  libmkl_vml_mc3.so.2
   14.0 MiB [#         ]  libmkl_gf_ilp64.so.2
   14.0 MiB [#         ]  libmkl_gf_ilp64.so
   13.9 MiB [#         ]  libmkl_vml_avx512.so.2
   13.1 MiB [#         ]  libomptarget.so
   12.2 MiB [#         ] /lib-dynload
   11.2 MiB [#         ]  libomptarget.rtl.x86_64.so
   10.8 MiB [#         ] /numpy
    8.5 MiB [#         ]  libmkl_vml_def.so.2
    8.3 MiB [#         ]  tlt
    7.8 MiB [#         ] /h5py
    7.8 MiB [#         ]  libomptarget.sycl.wrap.so
    7.5 MiB [#         ]  libmkl_vml_cmpt.so.2
    7.5 MiB [#         ]  libmkl_scalapack_lp64.so.2
    7.5 MiB [#         ]  libmkl_scalapack_lp64.so
    7.5 MiB [#         ]  libmkl_scalapack_ilp64.so.2
    7.5 MiB [#         ]  libmkl_scalapack_ilp64.so
    7.3 MiB [#         ]  libasan.so.6.0.0
    7.3 MiB [#         ]  libasan.so.6
    7.3 MiB [#         ]  libasan.so
    7.0 MiB [          ]  libtsan.so.0.0.0
...
```

Notably, the `numpy` and `pandas` folders are the same size - but 1.5 includes way more shared libraries. `libmkl` seems to be numpy-related, as it's a set of math libraries.

**Comment by rokm** [comment]
The build with poetry 1.1 is using PyPI `numpy` wheel, while the build with 1.5 is using Anaconda `numpy` package (and hence MKL). In the latter case, you can see the duplicated files that were originally symlinks (e.g., `libmkl_core.so`, which was originally a symlink to `libmkl_core.so.2`).

Or perhaps this applies to all packages; i.e., poetry 1.1 installed PyPI wheels, while 1.5 is installing Anaconda packages.

**Comment by hintron** [comment]
Ok, thanks! So then there are two compounding problems: 1) the anaconda package is being used, and it is likely bigger than the numpy wheel, and 2) the symlinks in the anaconda package are duplicated, making everything 2-3x as big as it would be. If 2 got solved, would 1 be a problem still?

I'm hoping there is a way in poetry to revert back to using the wheels instead of Anaconda packages, but if that's not possible, then it sounds like I can just hold out using poetry 1.1 until PyInstaller 6 comes out. What is the ETA for v6?

**Comment by rokm** [comment]
> Ok, thanks! So then there are two compounding problems: 1) the anaconda package is being used, and it is likely bigger than the numpy wheel, and 2) the symlinks in the anaconda package are duplicated, making everything 2-3x as big as it would be. If 2 got solved, would 1 be a problem still?

Using Anaconda's `numpy` would still result in a larger frozen application due to MKL (you can do an estimate from above listings, by eliminating all duplicates) - it is up to you to decide if that is a problem or not.

> I'm hoping there is a way in poetry to revert back to using the wheels instead of Anaconda packages, but if that's not possible, then it sounds like I can just hold out using poetry 1.1 until PyInstaller 6 comes out. What is the ETA for v6?

There is no ETA; although as far as I'm concerned, there is just one or two things that need to be sorted out before it should be ready.
